File: VoidMail/Views/Inbox/EmailRowView.swift
import SwiftUI

struct EmailRowView: View {
    let email: Email
    var accountColor: Color = .accentSkyBlue
    @State private var appeared = false

    var body: some View {
        HStack(alignment: .top, spacing: 14) {
            // Avatar + Account color bar (left side)
            ZStack(alignment: .topLeading) {
                InitialsAvatar(email.from.displayName, size: 44)

                // Account color indicator bar
                RoundedRectangle(cornerRadius: 2)
                    .fill(accountColor)
                    .frame(width: 3, height: 44)
                    .offset(x: -8)
            }

            // Content
            VStack(alignment: .leading, spacing: 4) {
                // Sender + Timestamp
                HStack {
                    Text(email.from.displayName)
                        .font(email.isRead ? Typo.body : Typo.headline)
                        .foregroundColor(.textPrimary)
                        .lineLimit(1)

                    Spacer()

                    Text(email.date.relativeFormatted)
                        .monoTimestamp()
                }

                // Subject
                Text(email.subject)
                    .font(email.isRead ? Typo.subhead : Typo.body)
                    .foregroundColor(email.isRead ? .textSecondary : .textPrimary)
                    .lineLimit(1)

                // Preview
                Text(email.snippet)
                    .font(Typo.subhead)
                    .foregroundColor(.textTertiary)
                    .lineLimit(2)
                    .lineSpacing(2)

                // Indicators
                HStack(spacing: 10) {
                    if email.isStarred {
                        Image(systemName: "star.fill")
                            .font(.system(size: 13))
                            .foregroundColor(.accentPink)
                    }
                    if !email.attachments.isEmpty {
                        HStack(spacing: 3) {
                            Image(systemName: "paperclip")
                                .font(.system(size: 13))
                            Text("\(email.attachments.count)")
                                .font(Typo.meta)
                        }
                        .foregroundColor(.textTertiary)
                    }
                    if email.aiSummary != nil {
                        Image(systemName: "sparkles")
                            .font(.system(size: 13))
                            .foregroundColor(.accentSkyBlue)
                    }
                    if email.isAIPriority {
                        HStack(spacing: 3) {
                            Image(systemName: "bolt.fill")
                                .font(.system(size: 11))
                            Text("AI PRIORITY")
                                .font(.system(size: 11, weight: .bold, design: .monospaced))
                        }
                        .foregroundColor(.accentPink)
                    }
                }
                .padding(.top, 2)
            }

            // Unread indicator bar on trailing edge
            if !email.isRead {
                RoundedRectangle(cornerRadius: 1.5)
                    .fill(Color.accentSkyBlue)
                    .frame(width: 3, height: 30)
            }
        }
        .padding(20)
        .background(Color.bgEmailRow)
        .cornerRadius(8)
        .padding(.horizontal, 16)
        .padding(.vertical, 4)
        .opacity(appeared ? 1 : 0)
        .offset(x: appeared ? 0 : 20)
        .onAppear {
            withAnimation(.spring(response: 0.4, dampingFraction: 0.85)) {
                appeared = true
            }
        }
    }
}

// MARK: - Date Formatting

extension Date {
    var relativeFormatted: String {
        let calendar = Calendar.current
        if calendar.isDateInToday(self) {
            let formatter = DateFormatter()
            formatter.timeStyle = .short
            return formatter.string(from: self)
        } else if calendar.isDateInYesterday(self) {
            return "Yesterday"
        } else if calendar.isDate(self, equalTo: Date(), toGranularity: .weekOfYear) {
            let formatter = DateFormatter()
            formatter.dateFormat = "EEE"
            return formatter.string(from: self)
        } else {
            let formatter = DateFormatter()
            formatter.dateFormat = "MMM d"
            return formatter.string(from: self)
        }
    }
}

File: lib/views/inbox/email_row_view.dart
import 'package:flutter/material.dart';
import '../../design_system/colors.dart';
import '../../design_system/typography.dart';
import '../../design_system/components.dart';
import '../../models/email.dart';

/// Email list row with swipe gestures, avatar, and indicators
class EmailRowView extends StatefulWidget {
  final Email email;
  final int index;
  final VoidCallback onTap;
  final VoidCallback onSwipeLeft; // Delete
  final VoidCallback onSwipeRight; // Toggle read

  const EmailRowView({
    super.key,
    required this.email,
    required this.index,
    required this.onTap,
    required this.onSwipeLeft,
    required this.onSwipeRight,
  });

  @override
  State<EmailRowView> createState() => _EmailRowViewState();
}

class _EmailRowViewState extends State<EmailRowView>
    with SingleTickerProviderStateMixin {
  double _dragOffset = 0;
  bool _isDragging = false;
  late AnimationController _entranceController;
  late Animation<double> _entranceAnimation;

  @override
  void initState() {
    super.initState();
    _entranceController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 400),
    );
    _entranceAnimation = CurvedAnimation(
      parent: _entranceController,
      curve: Curves.easeOutCubic,
    );

    // Staggered entrance
    Future.delayed(Duration(milliseconds: 50 * widget.index), () {
      if (mounted) _entranceController.forward();
    });
  }

  @override
  void dispose() {
    _entranceController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return FadeTransition(
      opacity: _entranceAnimation,
      child: SlideTransition(
        position: Tween<Offset>(
          begin: const Offset(0.05, 0),
          end: Offset.zero,
        ).animate(_entranceAnimation),
        child: GestureDetector(
          onHorizontalDragStart: (_) => setState(() => _isDragging = true),
          onHorizontalDragUpdate: (details) {
            setState(() {
              _dragOffset += details.primaryDelta ?? 0;
              _dragOffset = _dragOffset.clamp(-80.0, 80.0);
            });
          },
          onHorizontalDragEnd: (details) {
            if (_dragOffset > 80) {
              widget.onSwipeRight();
            } else if (_dragOffset < -80) {
              widget.onSwipeLeft();
            }
            setState(() {
              _dragOffset = 0;
              _isDragging = false;
            });
          },
          onTap: widget.onTap,
          child: Stack(
            children: [
              // Swipe action backgrounds
              _buildSwipeBackground(),

              // Email row content
              AnimatedContainer(
                duration: _isDragging
                    ? Duration.zero
                    : const Duration(milliseconds: 300),
                curve: Curves.elasticOut,
                transform: Matrix4.translationValues(_dragOffset, 0, 0),
                child: Padding(
                  padding: const EdgeInsets.symmetric(
                    horizontal: 16,
                    vertical: 4,
                  ),
                  child: Container(
                  padding: const EdgeInsets.all(20),
                  decoration: BoxDecoration(
                    color: VoidColors.bgEmailRow,
                    borderRadius: BorderRadius.circular(8),
                  ),
                  child: Row(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      // Avatar with account color bar
                      _buildAvatar(),
                      const SizedBox(width: 14),

                      // Email content
                      Expanded(
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            // Sender + timestamp row
                            Row(
                              children: [
                                if (!widget.email.isRead)
                                  const Padding(
                                    padding: EdgeInsets.only(right: 6),
                                    child: UnreadDot(size: 6),
                                  ),
                                Expanded(
                                  child: Text(
                                    widget.email.from.displayName,
                                    style: widget.email.isRead
                                        ? Typo.body.copyWith(
                                            color: VoidColors.textSecondary,
                                          )
                                        : Typo.headline.copyWith(fontSize: 16),
                                    maxLines: 1,
                                    overflow: TextOverflow.ellipsis,
                                  ),
                                ),
                                Text(
                                  widget.email.relativeDate,
                                  style: Typo.monoSmall,
                                ),
                              ],
                            ),
                            const SizedBox(height: 4),

                            // Subject
                            Text(
                              widget.email.subject,
                              style: Typo.body.copyWith(
                                fontSize: 15,
                                color: widget.email.isRead
                                    ? VoidColors.textSecondary
                                    : VoidColors.textPrimary,
                              ),
                              maxLines: 1,
                              overflow: TextOverflow.ellipsis,
                            ),
                            const SizedBox(height: 2),

                            // Snippet
                            Text(
                              widget.email.snippet,
                              style: Typo.subhead.copyWith(
                                fontSize: 14,
                                color: VoidColors.textTertiary,
                              ),
                              maxLines: 2,
                              overflow: TextOverflow.ellipsis,
                            ),

                            // Indicators row
                            if (_hasIndicators) ...[
                              const SizedBox(height: 6),
                              _buildIndicators(),
                            ],
                          ],
                        ),
                      ),
                    ],
                  ),
                ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  bool get _hasIndicators =>
      widget.email.isStarred ||
      widget.email.attachments.isNotEmpty ||
      widget.email.aiSummary != null ||
      widget.email.isAIPriority;

  Widget _buildAvatar() {
    return Stack(
      children: [
        InitialsAvatar(
          name: widget.email.from.displayName,
          size: 44,
        ),
        // Account color indicator (left edge bar)
        if (widget.email.accountEmail != null)
          Positioned(
            left: -8,
            top: 8,
            bottom: 8,
            child: Container(
              width: 3,
              decoration: BoxDecoration(
                color: widget.email.isRead
                    ? VoidColors.accentPink
                    : VoidColors.accentYellow,
                borderRadius: BorderRadius.circular(2),
              ),
            ),
          ),
      ],
    );
  }

  Widget _buildIndicators() {
    return Row(
      children: [
        if (widget.email.isStarred)
          const Padding(
            padding: EdgeInsets.only(right: 8),
            child: Icon(
              Icons.star,
              size: 14,
              color: VoidColors.accentYellow,
            ),
          ),
        if (widget.email.attachments.isNotEmpty)
          Padding(
            padding: const EdgeInsets.only(right: 8),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                const Icon(
                  Icons.attach_file,
                  size: 14,
                  color: VoidColors.textTertiary,
                ),
                Text(
                  '${widget.email.attachments.length}',
                  style: Typo.caption,
                ),
              ],
            ),
          ),
        if (widget.email.aiSummary != null)
          const Padding(
            padding: EdgeInsets.only(right: 8),
            child: Icon(
              Icons.auto_awesome,
              size: 14,
              color: VoidColors.accentSkyBlue,
            ),
          ),
        if (widget.email.isAIPriority)
          Container(
            padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
            decoration: BoxDecoration(
              color: VoidColors.accentPink.withValues(alpha: 0.15),
              borderRadius: BorderRadius.circular(4),
            ),
            child: Text(
              'AI PRIORITY',
              style: Typo.caption.copyWith(
                fontSize: 10,
                color: VoidColors.accentPink,
                fontWeight: FontWeight.w600,
                letterSpacing: 1,
              ),
            ),
          ),
      ],
    );
  }

  Widget _buildSwipeBackground() {
    return Positioned.fill(
      child: Row(
        children: [
          // Right swipe - toggle read (sky blue)
          Expanded(
            child: Container(
              color: _dragOffset > 0
                  ? VoidColors.accentSkyBlue
                  : Colors.transparent,
              alignment: Alignment.centerLeft,
              padding: const EdgeInsets.only(left: 24),
              child: _dragOffset > 30
                  ? const Icon(
                      Icons.mark_email_read,
                      color: Colors.white,
                      size: 24,
                    )
                  : null,
            ),
          ),
          // Left swipe - delete (pink)
          Expanded(
            child: Container(
              color: _dragOffset < 0
                  ? VoidColors.accentPink
                  : Colors.transparent,
              alignment: Alignment.centerRight,
              padding: const EdgeInsets.only(right: 24),
              child: _dragOffset < -30
                  ? const Icon(
                      Icons.delete,
                      color: Colors.white,
                      size: 24,
                    )
                  : null,
            ),
          ),
        ],
      ),
    );
  }
}
